src/write/util.rs

```diff
@@ -54,22 +59,74 @@ impl WritableBuffer for Vec<u8> {
     }
 
     #[inline]
-    fn reserve(&mut self, additional: usize) -> Result<(), ()> {
-        self.reserve(additional);
+    fn reserve(&mut self, size: usize) -> Result<(), ()> {
+        assert_eq!(self.len(), 0, "Buffer must be empty");
+        self.reserve(size);
         Ok(())
     }
 
     #[inline]
-    fn resize(&mut self, new_len: usize, value: u8) {
-        self.resize(new_len, value);
+    fn resize(&mut self, new_len: usize) {
+        debug_assert!(new_len >= self.len());
+        self.resize(new_len, 0);
     }
 
     #[inline]
     fn write_bytes(&mut self, val: &[u8]) {
+        debug_assert!(self.len() + val.len() <= self.capacity());
         self.extend_from_slice(val)
     }
 }
 
+/// A [`WritableBuffer`] that streams data to a [`Write`]r.
+///
+/// It is advisable to use a buffered writer like [`BufWriter`](std::io::BufWriter) instead of an
+/// unbuffered writer like [`File`](std::fs::File).
+#[derive(Debug)]
+pub struct StreamingBuffer<W> {
+    writer: W,
+    length: usize,
+}
+
+impl<W> StreamingBuffer<W> {
+    /// Create a new `WriteBuf` backed by the given writer.
+    pub fn new(writer: W) -> Self {
+        StreamingBuffer { writer, length: 0 }
+    }
+
+    /// Unwraps this [`WriteBuf`] giving back the original writer.
+    pub fn into_inner(self) -> W {
+        self.writer
+    }
+}
+
+impl<W: Write> WritableBuffer for StreamingBuffer<W> {
+    #[inline]
+    fn len(&self) -> usize {
+        self.length
+    }
+
+    #[inline]
+    fn reserve(&mut self, _additional: usize) -> Result<(), ()> {
+        Ok(())
+    }
+
+    #[inline]
+    fn resize(&mut self, new_len: usize) {
+        debug_assert!(new_len >= self.length);
+        while self.length < new_len {
+            let write_amt = (new_len - self.length - 1) % 1024 + 1;
+            self.write_bytes(&[0; 1024][..write_amt]);
+        }
+    }
+
+    #[inline]
+    fn write_bytes(&mut self, val: &[u8]) {
+        self.writer.write_all(val).unwrap();
```

Comment:
I'm not sure that unwrap is correct here.

Reply:
It isn't. I forgot to change this function to return `std::io::Result<()>`.

Reply:
It'd be nice if instead of returning a `std::io::Result` here, we cached the error value and ignored all future writes, and then returned the error value at the end.